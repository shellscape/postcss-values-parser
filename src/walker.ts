/*
  Copyright Â© 2025 Andrew Powell

  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  The above copyright notice and this permission notice shall be
  included in all copies or substantial portions of this Source Code Form.
*/

interface WalkCallback {
  (node: any, index: number): any;
}

const createWalker = (type: string) => {
  return function (this: any, callback: WalkCallback) {
    let index = 0;
    const walk = (node: any): any => {
      if (node.type === type.toLowerCase()) {
        const result = callback(node, index++);
        if (result === false) return false;
      }

      if (node.nodes && node.nodes.length > 0) {
        for (const child of node.nodes) {
          const result = walk(child);
          if (result === false) return false;
        }
      }
    };

    return walk(this);
  };
};

export const registerWalkers = (Container: any) => {
  const walkerTypes = [
    'Funcs',
    'Words',
    'Numerics',
    'Operators',
    'Quoteds',
    'UnicodeRanges',
    'Comments',
    'Punctuations'
  ];

  for (const walkerType of walkerTypes) {
    const methodName = `walk${walkerType}`;
    const nodeType = walkerType.toLowerCase().slice(0, -1); // Remove 's' and lowercase

    Container.prototype[methodName] = createWalker(nodeType);
  }

  // Special case for walkType
  Container.prototype.walkType = function (type: string, callback: WalkCallback) {
    return createWalker(type).call(this, callback);
  };
};
