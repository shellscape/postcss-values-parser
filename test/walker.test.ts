/*
  Copyright Â© 2025 Andrew Powell

  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  The above copyright notice and this permission notice shall be
  included in all copies or substantial portions of this Source Code Form.
*/
import { Container } from 'postcss';
import { describe, expect, it } from 'vitest';

import { parse, registerWalkers } from '../src/index.js';
import { values } from './fixtures/walker.json';

registerWalkers(Container);

describe('walker functionality', () => {
  for (const { value, walkers } of values) {
    const root = parse(value);

    for (const { length, type } of walkers) {
      it(`should walk ${type} in: ${value}`, () => {
        const nodes: any[] = [];

        (root as any)[`walk${type}`]((node: any) => {
          nodes.push(node);
        });

        expect(nodes.length).toBe(length);
        expect(nodes).toMatchSnapshot();
      });
    }
  }
});
