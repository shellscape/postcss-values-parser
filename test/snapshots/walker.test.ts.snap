// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`walker functionality > should walk Funcs in: calc((foo(768px - 100vw) / 2) - 15px) // batman
//joker 1`] = `
[
  {
    "inputs": [
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
    ],
    "isColor": false,
    "isVar": false,
    "name": "calc",
    "nodes": [
      {
        "nodes": [
          {
            "isColor": false,
            "isVar": false,
            "name": "foo",
            "nodes": [
              {
                "raws": {},
                "source": {
                  "end": {
                    "column": 16,
                    "line": 1,
                    "offset": 15,
                  },
                  "inputId": 0,
                  "start": {
                    "column": 11,
                    "line": 1,
                    "offset": 10,
                  },
                },
                "type": "numeric",
                "unit": "px",
                "value": "768",
              },
              {
                "raws": {},
                "source": {
                  "end": {
                    "column": 18,
                    "line": 1,
                    "offset": 17,
                  },
                  "inputId": 0,
                  "start": {
                    "column": 17,
                    "line": 1,
                    "offset": 16,
                  },
                },
                "type": "operator",
                "value": " - ",
              },
              {
                "raws": {},
                "source": {
                  "end": {
                    "column": 24,
                    "line": 1,
                    "offset": 23,
                  },
                  "inputId": 0,
                  "start": {
                    "column": 19,
                    "line": 1,
                    "offset": 18,
                  },
                },
                "type": "numeric",
                "unit": "vw",
                "value": "100",
              },
            ],
            "params": "",
            "raws": {},
            "source": {
              "end": {
                "column": 25,
                "line": 1,
                "offset": 24,
              },
              "inputId": 0,
              "start": {
                "column": 7,
                "line": 1,
                "offset": 6,
              },
            },
            "type": "func",
            "value": "",
          },
          {
            "raws": {},
            "source": {
              "end": {
                "column": 27,
                "line": 1,
                "offset": 26,
              },
              "inputId": 0,
              "start": {
                "column": 26,
                "line": 1,
                "offset": 25,
              },
            },
            "type": "operator",
            "value": "/",
          },
          {
            "raws": {},
            "source": {
              "end": {
                "column": 29,
                "line": 1,
                "offset": 28,
              },
              "inputId": 0,
              "start": {
                "column": 28,
                "line": 1,
                "offset": 27,
              },
            },
            "type": "numeric",
            "unit": "",
            "value": "2",
          },
        ],
        "raws": {},
        "source": {
          "end": {
            "column": 30,
            "line": 1,
            "offset": 29,
          },
          "inputId": 0,
          "start": {
            "column": 6,
            "line": 1,
            "offset": 5,
          },
        },
        "type": "parentheses",
        "value": "()",
      },
      {
        "raws": {},
        "source": {
          "end": {
            "column": 32,
            "line": 1,
            "offset": 31,
          },
          "inputId": 0,
          "start": {
            "column": 31,
            "line": 1,
            "offset": 30,
          },
        },
        "type": "operator",
        "value": " - ",
      },
      {
        "raws": {},
        "source": {
          "end": {
            "column": 37,
            "line": 1,
            "offset": 36,
          },
          "inputId": 0,
          "start": {
            "column": 33,
            "line": 1,
            "offset": 32,
          },
        },
        "type": "numeric",
        "unit": "px",
        "value": "15",
      },
    ],
    "params": "",
    "raws": {},
    "source": {
      "end": {
        "column": 38,
        "line": 1,
        "offset": 37,
      },
      "inputId": 0,
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": "func",
    "value": "",
  },
  {
    "inputs": [
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
    ],
    "isColor": false,
    "isVar": false,
    "name": "foo",
    "nodes": [
      {
        "raws": {},
        "source": {
          "end": {
            "column": 16,
            "line": 1,
            "offset": 15,
          },
          "inputId": 0,
          "start": {
            "column": 11,
            "line": 1,
            "offset": 10,
          },
        },
        "type": "numeric",
        "unit": "px",
        "value": "768",
      },
      {
        "raws": {},
        "source": {
          "end": {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
          "inputId": 0,
          "start": {
            "column": 17,
            "line": 1,
            "offset": 16,
          },
        },
        "type": "operator",
        "value": " - ",
      },
      {
        "raws": {},
        "source": {
          "end": {
            "column": 24,
            "line": 1,
            "offset": 23,
          },
          "inputId": 0,
          "start": {
            "column": 19,
            "line": 1,
            "offset": 18,
          },
        },
        "type": "numeric",
        "unit": "vw",
        "value": "100",
      },
    ],
    "params": "",
    "raws": {},
    "source": {
      "end": {
        "column": 25,
        "line": 1,
        "offset": 24,
      },
      "inputId": 0,
      "start": {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
    },
    "type": "func",
    "value": "",
  },
]
`;

exports[`walker functionality > should walk Numerics in: calc((foo(768px - 100vw) / 2) - 15px) // batman
//joker 1`] = `
[
  {
    "inputs": [
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
    ],
    "raws": {},
    "source": {
      "end": {
        "column": 16,
        "line": 1,
        "offset": 15,
      },
      "inputId": 0,
      "start": {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
    },
    "type": "numeric",
    "unit": "px",
    "value": "768",
  },
  {
    "inputs": [
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
    ],
    "raws": {},
    "source": {
      "end": {
        "column": 24,
        "line": 1,
        "offset": 23,
      },
      "inputId": 0,
      "start": {
        "column": 19,
        "line": 1,
        "offset": 18,
      },
    },
    "type": "numeric",
    "unit": "vw",
    "value": "100",
  },
  {
    "inputs": [
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
    ],
    "raws": {},
    "source": {
      "end": {
        "column": 29,
        "line": 1,
        "offset": 28,
      },
      "inputId": 0,
      "start": {
        "column": 28,
        "line": 1,
        "offset": 27,
      },
    },
    "type": "numeric",
    "unit": "",
    "value": "2",
  },
  {
    "inputs": [
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
    ],
    "raws": {},
    "source": {
      "end": {
        "column": 37,
        "line": 1,
        "offset": 36,
      },
      "inputId": 0,
      "start": {
        "column": 33,
        "line": 1,
        "offset": 32,
      },
    },
    "type": "numeric",
    "unit": "px",
    "value": "15",
  },
]
`;

exports[`walker functionality > should walk Words in: calc((foo(768px - 100vw) / 2) - 15px) // batman
//joker 1`] = `
[
  {
    "inputs": [
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
    ],
    "isColor": false,
    "isHex": false,
    "isUrl": false,
    "isVariable": false,
    "raws": {},
    "source": {
      "end": {
        "column": 48,
        "line": 1,
        "offset": 47,
      },
      "inputId": 0,
      "start": {
        "column": 42,
        "line": 1,
        "offset": 41,
      },
    },
    "type": "word",
    "value": "batman",
  },
  {
    "inputs": [
      {
        "css": "<unknown>",
        "hasBOM": false,
        "id": "<input css [ID]>",
      },
    ],
    "isColor": false,
    "isHex": false,
    "isUrl": false,
    "isVariable": false,
    "raws": {},
    "source": {
      "end": {
        "column": 8,
        "line": 2,
        "offset": 55,
      },
      "inputId": 0,
      "start": {
        "column": 3,
        "line": 2,
        "offset": 50,
      },
    },
    "type": "word",
    "value": "joker",
  },
]
`;
